lib/puma/minissl.rb

```diff
@@ -133,7 +133,7 @@ def write(data)
 
           return data_size if need == 0
 
-          data = data[wrote..-1]
+          data = data.byteslice(wrote..-1)
```

Comment:
@MSP-Greg Do you understand why 5.1.1 doesn't _seem_ to have the problem stated in #2542? This line of code wasn't changed in #2519 (but surrounding code was: https://github.com/puma/puma/pull/2519/files#diff-5f9e58f45420759185f3cff5203a523482b915cc3e331bb0f060977c2078d903)

Reply:
I will look at 5.1.1 and 5.2.0.  I believe the bug requires a response with multi-byte characters, longer than 16kB, and ssl.  And, since some html parsers can be rather forgiving...

Reply:
If I'm reading the code right, I think the bug was in changing line 133 from `data.bytesize` to `data`, no?

Reply:
Confused.  Do you mean changing `data.bytesize` to `data_size`, which is set to the value of `data.bytesize`?

Anyway, I'm working thru it.  The code in 5.1.1 was writing 512 byte chunks to the TCPSocket, which seemed not optimized.  But, it also appears that it wrote all the data with `@engine.write`, where the current code is writing it in 16kB chunks.  Because 5.1.1 was writing everything with `@engine.write`, maybe the issues with how it was looping weren't apparent?

Trying to figure out why the difference in write size is happening...  Regardless, I believe the current code is a lot faster?

Reply:
I had to review the code changes in minissl.c, but I now recall.  It's involved, but I'll write it up.